### PR TITLE
handling early terminal states

### PR DESCRIPTION
**Pull request by coverdrive**
both files have changes to handle early terminal states

### CODE REVIEW

File: rl/approximate_dynamic_programming.py [hunk 1]
```diff
@@ -167,17 +169,22 @@ def backward_evaluate(
...
+            return r + γ * (v[i-1].evaluate([s1]).item() if i > 0 and not
+                            mrp_f0_mu_triples[num_steps - i][0].is_terminal(s1)
+                            else 0.)
 
         v.append(
             approx0.solve(
                 [(s, mrp.transition_reward(s).expectation(return_))
-                 for s in mu.sample_n(num_state_samples)],
+                 for s in mu.sample_n(num_state_samples)
+                 if not mrp.is_terminal(s)],
```
Comment (TikhonJelvis) [inline rl/approximate_dynamic_programming.py:187]: With this approach, you might get a lot less than `n` states sampled here if there are a lot of terminal states. Could this be a problem?

Comment (coverdrive) [reply rl/approximate_dynamic_programming.py:187]: Yes, but I'm not sure there's a way around it if we are sampling states with a simulation trace. If we have an analytic form for the non-terminal states distribution (in some examples, we do), then we can sample directly from that known distribution of non-terminal states (for a given time step).

Comment (TikhonJelvis) [reply rl/approximate_dynamic_programming.py:187]: Okay, that makes sense. Sounds like there's no point worrying about it unless it really becomes a problem.

**Review by TikhonJelvis — APPROVED** [review]
Looks good.

**Pull request merged**